File: golang/stack.go
package main

type stack struct {
	length int
	top   *node
}

func (s *stack) push(item int) {
	s.length++
	node := &node{data: item}
	if s.top == nil {
		s.top = node
		return
	}
	node.next = s.top
	s.top = node
}

func (s *stack) pop() int {
	if s.length != 0 {

		s.length--
		result := s.top.data
		s.top = s.top.next
		return result
	}
	// should be err maybe
	return -1
}

func (s *stack) peek() int {
	return s.top.data
}

File: js/src/day1/Stack.ts
type node<T> = {
    value: T;
    next?: node<T>;
};
export default class Stack<T> {
    public length: number;

    public head?: node<T>;

    constructor() {
        this.head = undefined;
        this.length = 0;
    }

    push(item: T): void {
        const node: node<T> = { value: item };
        this.length++;
        if (!this.head) {
            this.head = node;
            return;
        }
        node.next = this.head;
        this.head = node;
    }

    pop(): T | undefined {
        const result = this.head?.value;

        if (this.head) {
            this.length--;
            this.head = this.head.next;
        }
        return result;
    }

    peek(): T | undefined {
        return this.head?.value;
    }
}
